src/basler_synced_cameras_node.cpp

```diff
@@ -0,0 +1,255 @@
+    {
+        int num_init = 0;
+        int num_master = 0;
+        int num_slave = 0;
+        int64_t master_clock_id = 0;
+        bool multiple_masters = false;
+        for (const auto& camera : cameras)
+        {
+            camera.GevIEEE1588DataSetLatch();
+            switch(camera.GevIEEE1588StatusLatched())
+            {
+            case Basler_GigECamera::GevIEEE1588StatusLatched_Master:
+            {
+                ++num_master;
+                break;
+            }
+            case Basler_GigECamera::GevIEEE1588StatusLatched_Slave:
+            {
+                int64_t master_id = camera.GevIEEE1588ParentClockId();
+                if (num_slave == 0)
+                {
+                    master_clock_id = master_id;
+                }
+                else if(master_id != master_clock_id)
+                {
+                    multiple_masters = true;
+                }
+                ++num_slave;
+                break;
+            }
+            default:
+            {
+                ++num_init;
+            }
+            }
+        }
+        if (num_slave == cameras.size() && !multiple_masters)
+        {
+            ROS_INFO_STREAM("All camera clocks are PTP slaves to master clock " << master_clock_id);
+            break;
+        }
+        else if(num_slave == cameras.size())
+        {
+            ROS_INFO("All camera clocks are PTP slaves, but multiple masters are present");
+        }
+        else
+        {
+            ROS_INFO("Camera PTP status: %d initializing, %d masters, %d slaves", num_init, num_master, num_slave);
+        }
+        r.sleep();
+        ros::spinOnce();
+    }
+}
+
+void waitForPTPClockSync(const std::vector<Pylon::CBaslerGigEInstantCamera>& cameras, int max_offset_ns, int offset_window_s)
+{
+    ROS_INFO("Waiting for clock offsets < %d ns over %d s...", max_offset_ns, offset_window_s);
+    bool currently_below = false;
+    ros::Time below_start;
+    std::vector<int64_t> current_offsets(cameras.size());
+    int64_t current_max_abs_offset = 0;
+
+    ros::Time last_output = ros::Time::now();
+    while (ros::ok())
+    {
+        for (size_t i = 0; i < cameras.size(); ++i)
+        {
+            cameras[i].GevIEEE1588DataSetLatch();
+            current_offsets[i] = cameras[i].GevIEEE1588OffsetFromMaster();
+            current_max_abs_offset = std::max(current_max_abs_offset, std::abs(current_offsets[i]));
+        }
+        if (current_max_abs_offset < max_offset_ns)
+        {
+            if (!currently_below)
+            {
+                currently_below = true;
+                below_start = ros::Time::now();
+            }
+            else if ((ros::Time::now() - below_start).toSec() >= offset_window_s)
+            {
+                break;
+            }
+        }
+        else
+        {
+            currently_below = false;
+        }
+        // Print out current offsets periodically
+        if ((ros::Time::now()  - last_output).toSec() >= 1)
```

Comment:
This should definitely be DEBUG level.  You may also be able to use THROTTLE to handle this nicely for you, at least if you can package it all up in one print statement.